spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java

```
@@ -0,0 +1,209 @@
+		super.addInterceptor(interceptor);
+		if (interceptor instanceof ExecutorChannelInterceptor) {
+			this.executorInterceptorsSize++;
+		}
+	}
+
+	@Override
+	public void addInterceptor(int index, ChannelInterceptor interceptor) {
+		super.addInterceptor(index, interceptor);
+		if (interceptor instanceof ExecutorChannelInterceptor) {
+			this.executorInterceptorsSize++;
+		}
+	}
+
+	@Override
+	public boolean removeInterceptor(ChannelInterceptor interceptor) {
+		boolean removed = super.removeInterceptor(interceptor);
+		if (removed && interceptor instanceof ExecutorChannelInterceptor) {
+			this.executorInterceptorsSize--;
+		}
+		return removed;
+	}
+
+	@Override
+	public ChannelInterceptor removeInterceptor(int index) {
+		ChannelInterceptor interceptor = super.removeInterceptor(index);
+		if (interceptor != null && interceptor instanceof ExecutorChannelInterceptor) {
+			this.executorInterceptorsSize--;
+		}
+		return interceptor;
+	}
+
+	@Override
+	public boolean hasExecutorInterceptors() {
+		return this.executorInterceptorsSize > 0;
+	}
+
+	protected class MessageHandlingTask implements Runnable {
+
+		private final MessageHandlingRunnable delegate;
+
+		public MessageHandlingTask(MessageHandlingRunnable task) {
+			this.delegate = task;
+		}
+
+		@Override
+		public void run() {
+			Message<?> message = this.delegate.getMessage();
+			MessageHandler messageHandler = this.delegate.getMessageHandler();
+			Assert.notNull(messageHandler, "'messageHandler' must not be null");
+			Deque<ExecutorChannelInterceptor> interceptorStack = null;
+			try {
+				if (executorInterceptorsSize > 0) {
+					interceptorStack = new ArrayDeque<ExecutorChannelInterceptor>();
+					message = applyBeforeHandle(message, interceptorStack);
+					if (message == null) {
+						return;
+					}
+				}
+				messageHandler.handleMessage(message);
+				if (!CollectionUtils.isEmpty(interceptorStack)) {
+					triggerAfterMessageHandled(message, null, interceptorStack);
```

Since we're eating exceptions in this method, perhaps it should return the last exception (after the while loop completes) so we can re-throw it here.